### PR TITLE
Redirect on graph click

### PR DESCRIPTION
**Pull request by ghost**
Graphs from dashboard module are allowed to be clickable, like a bar on bar-charts, etc.
The click redirects to listing view depending on data from graph data (filters, model referred by graph, etc.) and clicked element on graph.

### CODE REVIEW

**Comment by ghost** [comment]
TODO:
* [x] hide filter graph-query
* [x] pointer arrow

File: src/ralph/admin/templates/admin/partials/filters.html [hunk 1]
```diff
@@ -2,15 +2,19 @@
...
+        {% if is_popup %}
+          <input type="hidden" name="_popup" value="1" />
+        {% endif %}
+        {% for spec in cl.filter_specs %}
+          {% admin_list_filter cl spec %}
+        {% endfor %}
+      </div>
+      <input type="button" class="button" id="filter-button" value="Filter" />
+    {% else %}
+      This is a graph preview. Filters disabled.
```
Comment (mkurek) [inline src/ralph/admin/templates/admin/partials/filters.html]: shouldn't this be placed in some tag (`<p>` etc?)

File: src/ralph/admin/mixins.py [hunk 2]
```diff
@@ -211,7 +211,22 @@ def _check_absolute_url(self, cls, model):
...
 
 
-class RalphAdminMixin(RalphAutocompleteMixin):
+class DashboardChangelistMixin(object):
+
+    def _is_graph_preview_view(self, request):
+        return request.GET.get('graph-query', '')
+
+    def get_list_filter(self, request):
+        from ralph.dashboards.admin_filters import ByGraphFilter
```
Comment (ar4s) [inline src/ralph/admin/mixins.py]: it's necessary to do import in this line?

File: src/ralph/admin/templates/admin/partials/filters.html [hunk 3]
```diff
@@ -2,15 +2,19 @@
...
+        {% if is_popup %}
+          <input type="hidden" name="_popup" value="1" />
+        {% endif %}
+        {% for spec in cl.filter_specs %}
+          {% admin_list_filter cl spec %}
+        {% endfor %}
+      </div>
+      <input type="button" class="button" id="filter-button" value="Filter" />
+    {% else %}
+      This is a graph preview. Filters are disabled.
```
Comment (ar4s) [inline src/ralph/admin/templates/admin/partials/filters.html:17]: it's possible to display query or something?

Comment (mkurek) [inline src/ralph/admin/templates/admin/partials/filters.html:17]: shouldn't this be placed in some tag (`<p>` etc)?

File: src/ralph/admin/templates/admin/partials/filters.html (hunk 1 above)
Comment (ghost) [reply src/ralph/admin/templates/admin/partials/filters.html]: It's ok to me now..
@ar4s 
should I change it?

File: src/ralph/admin/mixins.py [hunk 4]
```diff
@@ -211,7 +211,22 @@ def _check_absolute_url(self, cls, model):
...
+        filters = super().get_list_filter(request) or []
+        is_graph_model = getattr(self.model, '_allow_in_dashboard', False)
+        if is_graph_model and ByGraphFilter not in filters:
+            filters.append(ByGraphFilter,)
```
Comment (mkurek) [inline src/ralph/admin/mixins.py]: this comma is not necessary (even misleading)

File: src/ralph/dashboards/admin_filters.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    def tokenize(self, query_value):
+        return query_value.split(self.sep) if self.sep in query_value else (
+            None, None
+        )
+
+    def queryset(self, request, queryset):
+        graph_pk, graph_item = self.tokenize(self.value() or '')
+        if graph_pk and graph_item:
+            graph = get_object_or_404(Graph, pk=graph_pk)
+            queryset = graph.apply_parital_filtering(queryset)
```
Comment (mkurek) [inline src/ralph/dashboards/admin_filters.py:29]: 👍

**Review by mkurek — CHANGES_REQUESTED** [review]

File: src/ralph/admin/templates/admin/partials/filters.html (hunk 3 above)
Comment (ghost) [reply src/ralph/admin/templates/admin/partials/filters.html:17]: It's ok to me now..
@ar4s 
should I change it?

File: src/ralph/admin/mixins.py [hunk 6]
```diff
@@ -273,6 +288,9 @@ def changelist_view(self, request, extra_context=None):
         for view in self.list_views:
             views.append(view)
         extra_context['list_views'] = views
+        extra_context['is_graph_preview_view'] = self._is_graph_preview_view(
```
Comment (ar4s) [inline src/ralph/admin/mixins.py]: can you move this to ``DashboardChangelistMixin``?

**Comment by ghost** [comment]
image

**Review by ar4s — CHANGES_REQUESTED** [review]

File: src/ralph/admin/mixins.py (hunk 2 above)
Comment (ghost) [reply src/ralph/admin/mixins.py]: cross imports

File: src/ralph/admin/templates/admin/partials/filters.html (hunk 3 above)
Comment (ghost) [reply src/ralph/admin/templates/admin/partials/filters.html:17]: we agreed to skip it this time

**Review by ar4s — APPROVED** [review]

**Review by mkurek — APPROVED** [review]

**Pull request merged**